init.lua

```diff
@@ -95,111 +95,139 @@ dofile(modpath.."/saplings.lua")
+	minetest.register_decoration(translate_biome_defs(moretrees.willow_biome, "willow"))
 end
 
 if moretrees.enable_birch then
-	biome_lib:register_generate_plant(moretrees.birch_biome, moretrees.spawn_birch_object)
+	minetest.register_decoration(translate_biome_defs(moretrees.birch_biome, "birch"))
 end
 
 if moretrees.enable_spruce then
-	biome_lib:register_generate_plant(moretrees.spruce_biome, moretrees.spawn_spruce_object)
+	minetest.register_decoration(translate_biome_defs(moretrees.spruce_biome, "spruce"))
 end
 
 if moretrees.enable_jungle_tree then
-	biome_lib:register_generate_plant(moretrees.jungletree_biome, moretrees.spawn_jungletree_object)
+	minetest.register_decoration(translate_biome_defs(moretrees.jungletree_biome, "jungletree"))
 end
 
 if moretrees.enable_fir then
-	biome_lib:register_generate_plant(moretrees.fir_biome, moretrees.spawn_fir_object)
+	minetest.register_decoration(translate_biome_defs(moretrees.fir_biome, "fir", 1))
 	if minetest.get_modpath("snow") then
-		biome_lib:register_generate_plant(moretrees.fir_biome_snow, moretrees.spawn_fir_snow_object)
+		minetest.register_decoration(translate_biome_defs(moretrees.fir_biome_snow, "fir", 2))
 	end
 end
 
 if moretrees.enable_poplar then
-	biome_lib:register_generate_plant(moretrees.poplar_biome, moretrees.spawn_poplar_object)
-	biome_lib:register_generate_plant(moretrees.poplar_biome_2, moretrees.spawn_poplar_object)
-	biome_lib:register_generate_plant(moretrees.poplar_biome_3, moretrees.spawn_poplar_object)
-	biome_lib:register_generate_plant(moretrees.poplar_small_biome, moretrees.spawn_poplar_small_object)
-	biome_lib:register_generate_plant(moretrees.poplar_small_biome_2, moretrees.spawn_poplar_small_object)
+	minetest.register_decoration(translate_biome_defs(moretrees.poplar_biome, "popular", 1))
+	minetest.register_decoration(translate_biome_defs(moretrees.poplar_biome_2, "popular", 2))
+	minetest.register_decoration(translate_biome_defs(moretrees.poplar_biome_3, "popular", 3))
+	minetest.register_decoration(translate_biome_defs(moretrees.poplar_small_biome, "popular", 4))
+	minetest.register_decoration(translate_biome_defs(moretrees.poplar_small_biome_2, "popular", 5))
+end
+
+for k, v in pairs(deco_ids) do
+	deco_ids[k] = minetest.get_decoration_id(v)
 end
+minetest.set_gen_notify("decoration", deco_ids)
+
+minetest.register_on_generated(function(minp, maxp, blockseed)
+    local g = minetest.get_mapgen_object("gennotify")
+	--minetest.chat_send_all(dump(g))
+    local locations = {}
+	for _, id in pairs(deco_ids) do
+		local deco_locations = g["decoration#" .. id] or {}
+		for _, pos in pairs(deco_locations) do
+			locations[#locations+1] = pos
+		end
+	end
+
+    if #locations == 0 then return end
+    for _, pos in ipairs(locations) do
+		--minetest.chat_send_all("yay")
+        local timer = minetest.get_node_timer({x=pos.x, y=pos.y+1, z=pos.z})
+        timer:start(math.random(2,10))
+		--minetest.set_node(pos, {name="default:stone"})
+    end
+end)
```

Comment:
`math.random` should be localised in on_generated callbacks

Reply:
doesnt need to be.

Reply:
further more, math.random is used elsewhere across this mod. if you wish to optimize this, should be done in another pr that hits all instances as once